### PR TITLE
Bidirectional Wrapper

### PR DESCRIPTION
**Pull request by farizrahman4u**
```python
model = Sequential()
model.add(Bidirectional(LSTM(10, return_sequences=True), input_shape=(10, 20)))
model.add(Bidirectional(GRU(5), merge_mode='sum'))
model.add(Activation('sigmoid'))
```

### CODE REVIEW

File: keras/layers/wrappers.py [hunk 1]
```diff
@@ -133,3 +133,99 @@ def step(x, states):
...
+            return self.forward.get_output_shape_for(input_shape)
+        elif self.merge_mode == 'concat':
+            shape = list(self.forward.get_output_shape_for(input_shape))
+            shape[-1] *= 2
+            return tuple(shape)
+
+    def call(self, X, mask=None):
+
+        def reverse(x):
+            rev = K.permute_dimensions(x, (1, 0, 2))[::-1]
```
Comment (fchollet) [inline keras/layers/wrappers.py]: This will break with TF, and it would be more efficient to use `go_backwards` since it avoid the back and forth with `permute_dimensions`.

Comment (farizrahman4u) [inline keras/layers/wrappers.py]: Currently that part of rnn is still buggy. For example in theano backend, with go_backwards=True and unroll=False, both the input and the mask is reversed, but if unroll=True, then only the input is reversed.

**Comment by fchollet** [comment]
Please group your commits together to avoid launching many Travis CI runs.

**Comment by farizrahman4u** [comment]
Sorry about the multiple commits.. on mobile device.

**Comment by farizrahman4u** [comment]
ping @fchollet

File: keras/layers/wrappers.py [hunk 2]
```diff
@@ -133,3 +133,92 @@ def step(x, states):
...
+    ```python
+    model = Sequential()
+    model.add(Bidirectional(LSTM(10, return_sequences=True), input_shape=(5, 10)))
+    model.add(Bidirectional(LSTM(10)))
+    model.add(Dense(5))
+    model.add(Activation('softmax'))
+    model.compile(loss='categorical_crossentropy', optimizer='rmsprop')
+    ```
+    '''
+    def __init__(self, layer, merge_mode='concat', weights=None, **kwargs):
```
Comment (fchollet) [inline keras/layers/wrappers.py]: The `merge_mode` argument is never validated. Additionally, we should consider a `None` mode that just makes the layer return `(forward, backward)`

File: keras/layers/wrappers.py [hunk 3]
```diff
@@ -133,3 +133,92 @@ def step(x, states):
...
+        self.forward = layer
+        config = layer.get_config()
+        config['go_backwards'] = not config['go_backwards']
+        self.reverse = layer.__class__.from_config(config)
```
Comment (fchollet) [inline keras/layers/wrappers.py]: `self.forward / self.reverse` is inconsistent, it should be `self.forward / self.backwards`

Comment (fchollet) [inline keras/layers/wrappers.py]: Actually, maybe `self.forward_layer`, `self.backward_layer` to be extra explicit.

File: keras/layers/wrappers.py [hunk 4]
```diff
@@ -133,3 +133,92 @@ def step(x, states):
...
+    def reset_states(self):
+        self.forward.reset_states()
+        self.reverse.reset_states()
+
+    def build(self, input_shape):
+        self.forward.build(input_shape)
+        self.reverse.build(input_shape)
+        params = ['trainable_weights', 'non_trainable_weights', 'updates', 'regularizers']
+        for p in params:
+            setattr(self, p, getattr(self.forward, p, []) + getattr(self.reverse, p, []))
```
Comment (fchollet) [inline keras/layers/wrappers.py]: Let's rather implement a `@property` for these attributes. Much safer and more readable.

File: keras/layers/wrappers.py [hunk 5]
```diff
@@ -133,3 +133,92 @@ def step(x, states):
...
+        self.constraints = {}
+        if hasattr(self.forward, 'constraints'):
+            self.constraints.update(self.forward.constraints)
+            self.constraints.update(self.reverse.constraints)
```
Comment (fchollet) [inline keras/layers/wrappers.py]: Same here.

**Comment by farizrahman4u** [comment]
@fchollet All green.

File: keras/engine/topology.py [hunk 6]
```diff
@@ -284,10 +284,14 @@ def __init__(self, **kwargs):
...
-        self.regularizers = []
-        self.constraints = {}  # dict {tensor: constraint instance}
+        if not hasattr(self, 'trainable_weights'):
+                self.trainable_weights = []
+        if not hasattr(self, 'non_trainable_weights'):
+                self.non_trainable_weights = []
+        if not hasattr(self, 'regularizers'):
+                self.regularizers = []
+        if not hasattr(self, 'constraints'):
+                self.constraints = {}  # dict {tensor: constraint instance}
```
Comment (fchollet) [inline keras/engine/topology.py]: Better to leave this part of the code unchanged and implement a property setter on the wrapper.

Comment (farizrahman4u) [inline keras/engine/topology.py]: So when these properties are set, the lists should be split to 2 and set to forward and backward rnns separately?
For e.g, what happens when I do bi_rnn.trainable_weights=[a,b,c,d]? Set [a,b] to forward rnn and [c, d] to backward rnn?
In that case what about constrains? Its a dict, the order will be messed up.

Comment (fchollet) [inline keras/engine/topology.py]: That's a good point. Just fix the indentation in that part of the code, then.

Comment (farizrahman4u) [inline keras/engine/topology.py]: No pep8 errors. Can you comment on the particular line?

Comment (fchollet) [inline keras/engine/topology.py]: It seems you are indenting with 8 spaces. It should be 4.

**Comment by farizrahman4u** [comment]
Done.

**Comment by farizrahman4u** [comment]
And let me know what other wrappers you need. Highway, residual.. ?

**Comment by fchollet** [comment]
> And let me know what other wrappers you need. Highway, residual.. ?

I don't think it makes sense to have these anymore, since:
- they typically wrap blocks of layers, not a single layer
- they are really easy to set up with the functional API

If I think of something I'll let you know.

File: examples/imdb_bidirectional_lstm.py [hunk 7]
```diff
@@ -10,7 +10,7 @@
 
 from keras.preprocessing import sequence
 from keras.models import Model
-from keras.layers import Dense, Dropout, Embedding, LSTM, Input, merge
+from keras.layers import Dense, Dropout, Embedding, LSTM, Input, merge, Bidirectional
```
Comment (fchollet) [inline examples/imdb_bidirectional_lstm.py]: Two unused imports (merge, Model)

**Comment by fchollet** [comment]
LGTM, thanks!

**Comment by fchollet** [comment]
Ad a post-mortem: you've consumed ~28 hours of Travis CI with this PR. Please group your commits next time; the typical PR shouldn't require more than ~3 test runs.

**Comment by farizrahman4u** [comment]
Cool. Next time :+1:

**Comment by fchollet** [comment]
@farizrahman4u it looks like Theano fails when the output dim of the RNN is 1. Unclear why. Can you look into it?

**Comment by PiranjaF** [comment]
This is amazing. Nice job!

**Comment by farizrahman4u** [comment]
I think its a bug in some theano optimization.. if a dim is 1 it gets automatically squeezed out(under certain conditions). Probably they forgot to add back the 1 dim. Can you try it with all theano optimizations turned off?

**Comment by fchollet** [comment]
I think it's most likely a bug with the reverse op you added in the case
where the output dim is 1.

**Comment by farizrahman4u** [comment]
No, this has nothing to do with this pull request. I will start a new issue where this can be investigated.

**Comment by farizrahman4u** [comment]
@fchollet #3515

**Comment by farizrahman4u** [comment]
@fchollet, Found it. The bug is in theano's scan op.

**Comment by fchollet** [comment]
Cool, good to know. Thanks for looking into it.

**Comment by vinayakumarr** [comment]
imdb bidirectional example is not working. please check it #3698

**Comment by pifelbo** [comment]
See https://github.com/fchollet/keras/issues/4241 on how the Bidirectional wrapper does not play nice with the `layer.trainable` attribute.
